src/mm/memory_manager.rs

```diff
@@ -20,46 +33,64 @@ use crate::util::heap::layout::vm_layout_constants::HEAP_START;
-// A VM binding can borrow a mutable reference directly from Box<Mutator>, and call alloc(). Alternatively,
-// it can turn the Box pointer to a native pointer (*mut Mutator), then forge a mut reference from the native
-// pointer. In either way, the VM binding code needs to guarantee the safety.
-
-// How the VM gets mutator/collector/tracelocal handles:
-// * Mutator: from bind_mutator() as Box<Mutator>
-// * Collector: from Collection::spawn_worker_thread() as &mut Collector
-// * TraceLocal: Scanning::* as &mut TraceLocal
-
+/// Run the main loop for the GC controller thread. This method does not return.
+///
+/// Arguments:
+/// * `mmtk`: The reference to an MMTk instance.
+/// * `tls`: The thread that will be used as the GC controller.
 pub fn start_control_collector<VM: VMBinding>(mmtk: &MMTK<VM>, tls: OpaquePointer) {
     mmtk.plan.base().control_collector_context.run(tls);
 }
 
+/// Initializes an MMTk instance. A VM should call this method after creating an [MMTK](../mmtk/struct.MMTK.html)
+/// instance and before using any of the methods provided in MMTk.
+///
+/// Arguments:
+/// * `mmtk`: The reference to an MMTk instance to initialize.
+/// * `heap_size`: The heap size for the MMTk instance in bytes.
 pub fn gc_init<VM: VMBinding>(mmtk: &'static mut MMTK<VM>, heap_size: usize) {
     crate::util::logger::init().unwrap();
     mmtk.plan.gc_init(heap_size, &mmtk.vm_map, &mmtk.scheduler);
 }
 
+/// Request MMTk to create a mutator for the given thread. For performance reasons, A VM should
+/// store the returned mutator in a thread local storage that can be accessed efficiently.
+///
+/// Arguments:
+/// * `mmtk`: The reference to an MMTk instance.
+/// * `tls`: The thread that will be associated with the mutator.
 pub fn bind_mutator<VM: VMBinding>(
     mmtk: &'static MMTK<VM>,
     tls: OpaquePointer,
 ) -> Box<Mutator<SelectedPlan<VM>>> {
     SelectedPlan::bind_mutator(&mmtk.plan, tls, mmtk)
 }
 
+/// Reclaim a mutator that is no longer needed.
+///
+/// Arguments:
+/// * `mutator`: The reference to the mutator to be destroyed.
 pub fn destroy_mutator<VM: VMBinding>(mutator: Box<Mutator<SelectedPlan<VM>>>) {
     drop(mutator);
 }
 
+/// Flush the mutator's local states.
+///
+/// Arguments:
+/// * `mutator`: The reference to the mutator.
 pub fn flush_mutator<VM: VMBinding>(mutator: &mut Mutator<SelectedPlan<VM>>) {
     mutator.flush()
 }
 
+/// Allocate memory for an object. For performance reasons, a VM should
+/// implement the allocation fast-path on their side rather than just calling this function.
+///
+/// Arguments:
+/// * `mutator`: The mutator to perform this allocation request.
+/// * `size`: The number of bytes required for the object.
+/// * `align`: Required alignment for the object.
+/// * `offset`: Offset associated with the alignment.
+/// * `allocator`: The allocation semantic required for the allocation.
```

Comment:
If I remember correctly, we should use `semantics` instead of `semantic` in most cases, because the `Allocator` type includes more than one semantic.
In the rest of this file, `semantics` and `semantic` are both used.

Reply:
Yeah.

Reply:
But we won't have multiple semantics for a while.